File: typescript_test_crate/tests/14_system_return_default.ts

// NOTE: Default return value syntax (method(): type = default) not yet implemented.
// This test validates behavior when handler doesn't set system.return.


class SystemReturnDefaultTest {
    private _state: string;
    private _state_stack: Array<any>;
    private _state_context: Record<string, any>;
    private _return_value: any;

    constructor() {
        this._state_stack = [];
        this._state_context = {  };
        this._return_value = null;
        this._state = "Start";
        this._enter();
    }

    private _transition(target_state: string, exit_args: any = null, enter_args: any = null) {
        if (exit_args) {
            this._exit(...exit_args);
        } else {
            this._exit();
        }
        this._state = target_state;
        if (enter_args) {
            this._enter(...enter_args);
        } else {
            this._enter();
        }
    }

    private _change_state(target_state: string) {
        this._state = target_state;
    }

    private _dispatch_event(event: string, ...args: any[]) {
        const handler_name = `_s_${this._state}_${event}`;
        const handler = (this as any)[handler_name];
        if (handler) {
            return handler.apply(this, args);
        }
    }

    private _enter(...args: any[]) {
        if (this._state === "Start") {
            this._state_context["count"] = 0;
        }
    }

    private _exit(...args: any[]) {
        // No exit handlers
    }

    public handler_sets_value(): string {
        this._return_value = null
        this._dispatch_event("handler_sets_value")
        return this._return_value
    }

    public handler_no_return(): string | null {
        this._return_value = null
        this._dispatch_event("handler_no_return")
        return this._return_value
    }

    public get_count(): number {
        this._return_value = null
        this._dispatch_event("get_count")
        return this._return_value
    }

    private _s_Start_handler_sets_value() {
        this._return_value = "set_by_handler";
        return this._return_value;;
    }

    private _s_Start_handler_no_return() {
        // Does not set return - should return null/undefined
        this._state_context["count"] = this._state_context["count"] + 1;
    }

    private _s_Start_get_count() {
        this._return_value = this._state_context["count"];
        return this._return_value;;
    }
}


function main() {
    console.log("=== Test 14: System Return Default Behavior ===");
    const s = new SystemReturnDefaultTest();

    // Test 1: Handler explicitly sets return value
    const result1 = s.handler_sets_value();
    if (result1 !== "set_by_handler") {
        throw new Error(`Expected 'set_by_handler', got '${result1}'`);
    }
    console.log(`1. handler_sets_value() = '${result1}'`);

    // Test 2: Handler does NOT set return - should return null/undefined
    const result2 = s.handler_no_return();
    if (result2 !== null && result2 !== undefined) {
        throw new Error(`Expected null/undefined, got '${result2}'`);
    }
    console.log(`2. handler_no_return() = ${result2}`);

    // Test 3: Verify handler was called (side effect check)
    let count = s.get_count();
    if (count !== 1) {
        throw new Error(`Expected count=1, got ${count}`);
    }
    console.log(`3. Handler was called, count = ${count}`);

    // Test 4: Call again to verify idempotence
    const result4 = s.handler_no_return();
    if (result4 !== null && result4 !== undefined) {
        throw new Error(`Expected null/undefined again, got '${result4}'`);
    }
    count = s.get_count();
    if (count !== 2) {
        throw new Error(`Expected count=2, got ${count}`);
    }
    console.log(`4. Second call: result=${result4}, count=${count}`);

    console.log("PASS: System return default behavior works correctly");
}

main();

File: python_test_crate/tests/14_system_return_default.py
from typing import Any, Optional, List, Dict, Callable

class SystemReturnDefaultTest:
    def __init__(self):
        self._state_stack = []
        self._state_context = {}
        self._return_value = None
        self._state = "Start"
        self._enter()

    def _transition(self, target_state, exit_args = None, enter_args = None):
        if exit_args:
            self._exit(*exit_args)
        else:
            self._exit()
        self._state = target_state
        if enter_args:
            self._enter(*enter_args)
        else:
            self._enter()

    def _change_state(self, target_state):
        self._state = target_state

    def _dispatch_event(self, event, *args):
        handler_name = f"_s_{self._state}_{event}"
        handler = getattr(self, handler_name, None)
        if handler:
            return handler(*args)

    def _enter(self, *args):
        if self._state == "Start":
            self._state_context["count"] = 0


    def _exit(self, *args):
        # No exit handlers
        pass

    def handler_sets_value(self) -> str:
        self._return_value = None
        self._dispatch_event("handler_sets_value")
        return self._return_value

    def handler_no_return(self) -> str:
        self._return_value = None
        self._dispatch_event("handler_no_return")
        return self._return_value

    def get_count(self) -> int:
        self._return_value = None
        self._dispatch_event("get_count")
        return self._return_value

    def _s_Start_handler_sets_value(self) -> str:
        self._return_value = "set_by_handler"
        return

    def _s_Start_get_count(self) -> int:
        self._return_value = self._state_context["count"]
        return

    def _s_Start_handler_no_return(self) -> str:
        # Does not set return - should return None
        self._state_context["count"] = self._state_context["count"] + 1


def main():
    print("=== Test 14: System Return Default Behavior ===")
    s = SystemReturnDefaultTest()

    # Test 1: Handler explicitly sets return value
    result = s.handler_sets_value()
    assert result == "set_by_handler", f"Expected 'set_by_handler', got '{result}'"
    print(f"1. handler_sets_value() = '{result}'")

    # Test 2: Handler does NOT set return - should return None
    result = s.handler_no_return()
    assert result is None, f"Expected None, got '{result}'"
    print(f"2. handler_no_return() = {result}")

    # Test 3: Verify handler was called (side effect check)
    count = s.get_count()
    assert count == 1, f"Expected count=1, got {count}"
    print(f"3. Handler was called, count = {count}")

    # Test 4: Call again to verify idempotence
    result = s.handler_no_return()
    assert result is None, f"Expected None again, got '{result}'"
    count = s.get_count()
    assert count == 2, f"Expected count=2, got {count}"
    print(f"4. Second call: result={result}, count={count}")

    print("PASS: System return default behavior works correctly")

if __name__ == '__main__':
    main()

